### DISCUSSION TITLE
Is it possible to use MOOSE to optimize a variable?

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#26135] [Q&A General] [answered]
Say, there exists a simulation which outputs data and you have nearly similar data you want to compare it to, does MOOSE have any built in functionality to optimize a parameter, or multiple parameters (e.g material properties) to keep re-running the simulations until the outputs are close to the existing data?

That is, functionality to:

-provide an initial guess for the parameters to be optimized

-compare the error between the output data of the MOOSE simulation (from a PostProcessor) and the existing data, possibly restart the simulation early if the errors diverge too much in the beginning. For comparing transient simulations, it would be nice to have functionality to also tell it when to begin comparing.

-continue to optimize the parameters until the error is below a defined tolerance.

-output the optimized parameters.

An added bonus would be the ability to optimize the desired parameters/material properties over multiple pairs of simulations (same equations for all, just different setups) through some kind of least squares fitting, getting back a single value for each of those parameters that's the closest fit for all pairs.

I felt like this was a huge ask just writing it, so I was also wondering if there was any way I could make MOOSE interact directly with python (so I could implement this myself) without using a bash script to jump back and forth.

**Comment by ABallisat** [comment]
Interacting MOOSE with Python is surprisingly easy. You can use the SEACAS library (https://github.com/sandialabs/seacas) which has an EXODUSII reader in it for Python. We use this to read data into Python for analysis and doing exactly this (design space mapping, optimisation, sensitivity analysis, Monte Carlo integrals, etc.). I can't comment on doing this directly in MOOSE but its quite easy to do in Python. I run system calls to MOOSE from Python scripts, as MOOSE input files are just text files its trivial to automate writing these in Python with different input variables, or set them on the command line and automate the call generation. Depending on what hardware you are running on (laptop/desktop vs HPC cluster) you will want to choose your parallelism appropriately. If you are going on a cluster its not too hard to set up parallel execution of models. There are different approaches to this, e.g. Python's built in Multiprocessing or MPI. We have internal libraries for using the latter on clusters to run arbitrary numbers in parallel. All in all, if you can get the results into Python using SEACAS then you can do what you like!

**Reply by richmondodufisan** [reply]
Thanks for sharing, an exodus reader would be very useful!

>  I run system calls to MOOSE from Python scripts, as MOOSE input files are just text files its trivial to automate writing these in Python with different input variables, or set them on the command line and automate the call generation.

Could you expand a little bit more on this? What do you mean by running system calls to MOOSE and can you give an example? And what do you mean by writing MOOSE input files in python? Do you not use .i files?

**Reply by ABallisat** [reply, marked as answer]
use the os module in python, e.g.:
```python
import os
os.system("mpirun -np 4  myapp-opt -i myinputfile.i")
```
as for the input files they are just text files with a .i extension so something along the lines of:
```python
with open("myinputfile.i", w) as f:
    f.write("[my input file as a string]")
```
and then format the string with the values of your variables

**Reply by richmondodufisan** [reply]
Ahh I think I get what you're doing now. Yeah that makes a lot of sense. Thanks!

**Comment by josebastiase** [comment]
Hi,
You can do inverse modeling to fit parameters with isopod. Is based in [pest](https://pesthomepage.org/)

https://github.com/idaholab/isopod

**Reply by GiudGiud** [reply]
Much of isopod was merged into the optimization module too. Things like material property optimization are in the module

**Reply by richmondodufisan** [reply]
Interesting! I'll look into that thanks for sharing

**Reply by lindsayad** [reply]
@lynnmunday @maxnezdyur

**Reply by lynnmunday** [reply]
@richmondodufisan Most of what you are wanting to do would be covered by the optimization module:   https://mooseframework.inl.gov/moose/modules/optimization/
We mostly focused on gradient based optimization using the adjoint method.  It can handle transient and steady state problems although transient problems are more difficult.  There are several steady state examples in the test and examples directory of the module.  Here is an example for transient material inversion that hasn't been merged yet:
https://github.com/idaholab/moose/pull/25668
